### PR TITLE
Implement database driver mapping

### PR DESCRIPTION
**Pull request by ironashram**
This adds the same mapping capabilities of the YAML driver to the database driver.
The database structure now contains more fields to allow different authentication type for either upstream and downstream, also private keys and authorized keys

Please comment if there are changes required before getting this merged

### CODE REVIEW

File: sshpiperd/upstream/database/handler.go [hunk 1]
```diff
@@ -2,86 +2,313 @@ package database
...
+	return pipe, nil
+}
 
-	c, err := upstreamprovider.DialForSSH(addr)
+func (p *plugin) GetBytes(inputString string) ([]byte, error) {
 
-	if err != nil {
-		return nil, nil, err
+	if inputString != "" {
+		return []byte(inputString), nil
```
Comment (tg123) [inline sshpiperd/upstream/database/handler.go]: is this just covert string to []byte?

File: sshpiperd/upstream/database/handler.go [hunk 2]
```diff
@@ -2,86 +2,313 @@ package database
...
+		}
+		a, err := p.createAuthPipe(pipe, conn, challengeContext)
+		if err != nil {
+			return nil, nil, err
+		}
+		return c, a, nil
 	}
-	return c, &pipe, nil
+
+	return nil, nil, fmt.Errorf("username not [%v] found", pipe.Username)
```
Comment (tg123) [inline sshpiperd/upstream/database/handler.go]: empty "" not found? or it cannot be empty?

File: sshpiperd/upstream/database/handler.go [hunk 3]
```diff
@@ -2,86 +2,313 @@ package database
...
+		}
+
+		a.PublicKeyCallback = func(conn ssh.ConnMetadata, key ssh.PublicKey) (ssh.AuthPipeType, ssh.AuthMethod, error) {
+			return to(key)
+		}
+
+		return a, nil
+
+	default:
+		logger.Printf("unsupport type [%v], ignore section", pipe.FromType)
```
Comment (tg123) [inline sshpiperd/upstream/database/handler.go]: return err?

File: sshpiperd/upstream/database/handler.go [hunk 4]
```diff
@@ -2,86 +2,313 @@ package database
...
+
+		case authMapTypePassword:
+			toPass, err := p.GetBytes(pipe.ToPassword)
+
+			if err != nil {
+				return ssh.AuthPipeTypeDiscard, nil, err
+			}
 
-		hostKeyCallback = ssh.FixedHostKey(key)
+			return ssh.AuthPipeTypeMap, ssh.Password(string(toPass)), nil
```
Comment (tg123) [inline sshpiperd/upstream/database/handler.go]: why tobyte then to string again?

**Review by tg123 — COMMENTED** [review]
Thanks!

Any idea about mapping password to key vice verse
Yaml supports `From` and `To` object

File: sshpiperd/upstream/database/handler.go (hunk 1 above)
Comment (ironashram) [reply sshpiperd/upstream/database/handler.go]: removed unnecessary conversion function

File: sshpiperd/upstream/database/handler.go (hunk 2 above)
Comment (ironashram) [reply sshpiperd/upstream/database/handler.go]: should not be empty, modified

File: sshpiperd/upstream/database/handler.go (hunk 3 above)
Comment (ironashram) [reply sshpiperd/upstream/database/handler.go]: fixed

File: sshpiperd/upstream/database/handler.go (hunk 4 above)
Comment (ironashram) [reply sshpiperd/upstream/database/handler.go]: removed also this unnecessary conversion

**Comment by ironashram** [comment]
> Thanks!
> 
> Any idea about mapping password to key vice verse
> Yaml supports `From` and `To` object

regarding the mapping, well with this patch you can do that, you just need to specify the correct auth_map_type for upstream and downstram in the database and add the necessary keys, password and so on, i have a simple docker compose demo if you want i can share it

**Comment by tg123** [comment]
looking forward to see your demo

**Pull request merged**